market_analysis/views/default.py

```diff
@@ -0,0 +1,450 @@
+            msg = request.matchdict['name'] + ' was added to your portfolio.'
+        else:
+            msg = request.matchdict['name'] + ' is already in your portfolio.'
+        return {'msg': msg}
+
+
+@view_config(route_name='delete', renderer='../templates/delete_page.jinja2',
+             permission='secret')
+def delete_stock_from_portfolio(request):
+    if request.method == 'POST':
+        current_user_id = request.dbsession.query(Users).filter(
+                Users.username == request.authenticated_userid
+            ).first().id
+        new_stock_sym = request.matchdict['sym']
+        try:
+            query = request.dbsession.query(Stocks).\
+             filter(Stocks.symbol == new_stock_sym).first()
+            query_del = request.dbsession.query(Association)\
+                .filter(and_(Association.stock_id == query.id,
+                             Association.user_id == current_user_id)).first()
+            request.dbsession.delete(query_del)
+            msg = request.matchdict['sym'] + ' was removed from'\
+                ' your portfolio.'
+
+        except (AttributeError, UnmappedInstanceError):
+            msg = 'Failed: tried to remove a stock that is not in the'\
+                ' portfolio.'
+    else:
+        msg = 'Failed: improper request.'
+    return {'msg': msg}
+
+
+@view_config(route_name='portfolio', renderer="../templates/portfolio.jinja2",
+             permission='secret')
+def portfolio(request):
+    """
+    The main user portfolio page, displays a list of
+    their stocks and a graph. Update number of shares.
+    """
+    current_user_id = request.dbsession.query(Users).filter(
+                Users.username == request.authenticated_userid
+            ).first().id
+    query = request.dbsession.query(Users).\
+        filter(Users.id == current_user_id).first()
+    query = query.children
+
+    if request.method == 'POST':
+        updated_amount = request.POST['amount']
+        for item, val in request.POST.items():
+            if val == 'Update':
+                updated_stock = item
+        current_stock_id = request.dbsession.query(Stocks)\
+            .filter(Stocks.symbol == updated_stock).first().id
+        request.dbsession.query(Association)\
+            .filter(and_(current_stock_id == Association.stock_id,
+                    Association.user_id == current_user_id))\
+            .update({Association.shares: updated_amount})
+
+    list_of_stock_ids = []
+    for row in query:
+        list_of_stock_ids.append(row.child.symbol)
+    if len(list_of_stock_ids) == 0:
+        return HTTPFound(location=request.route_url('search'))
+    else:
+        elements = []
+        for stock in list_of_stock_ids:
+            elements.append({'Symbol': str(stock),
+                             'Type': 'price', 'Params': ['c']})
+        return build_graph(request, elements, True)
+
+
+def package_data(info, entries, msg):
+    """Package data for single_stock_details."""
```

Comment:
This docstring might be a good place to document what you expect the data you got from the API to look like.